### PR TITLE
[Enhance] Accelerate analyze result

### LINKED ISSUE #7631: run the tools/analysis_tools/analyze_results.py is so slow ,Is that normal？i feel it caculates with cpu.
```
(swin-t) E:\Object-Detection\Github\radar-detection>python tools/analysis_tools/analyze_results.py checkpoints/tood/devkit3_1calss_e500_af_r50_dcn_db/radar_tood.py checkpoints/tood/devkit3_1calss_e500_af_r50_dcn_db/pkl_and_cm/tood_95.pkl checkpoints/tood/devkit3_1calss_e500_af_r50_dcn_db/pkl_and_cm --show-score-thr 0.7
loading annotations into memory...
Done (t=0.00s)
creating index...
index created!
[>>>>>>>                                           ] 4/28, 0.0 task/s, elapsed: 209s, ETA:  1256s
```

### PR DESCRIPTION
**Pull request by chhluo**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Fix #7631

The time spent on [`eval_fn`](https://github.com/open-mmlab/mmdetection/blob/73b4e65a6a30435ef6a35f405e3474a4d9cfb234/tools/analysis_tools/analyze_results.py#L130) costs 1.5s, and on ['dataset.prepare_train_img(i)'](https://github.com/open-mmlab/mmdetection/blob/73b4e65a6a30435ef6a35f405e3474a4d9cfb234/tools/analysis_tools/analyze_results.py#L129) costs 0.01~0.05s. It is clear that we should accelerate the `eval_fn`. There are three nested loops in eval_fn: a loop for iou_thresh(0.5 ~ 0.95), a nested loop for classes, a nested loop for images. And the multi-processes design for fptp computing of list of images does not have expired effect when only one image passed in one time. The multi-proccesses design is unnecessary in this condition (one image a time).  But multi-proccesses design to compute mAP with different iou_thresh is necessary and effective.

image

After optimization, the time spend on `eval_fn` is reduced to less then one fifteenth of the original when `nproc = 4` (which is set by defualt). The evaluation time on results generated by retinanet_r50_fpn_1x_coco takes 10~15 mins.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet/core/evaluation/mean_ap.py [hunk 1]
```diff
@@ -611,13 +617,22 @@ def eval_map(det_results,
...
+                    [iou_thr for _ in range(num_imgs)],
+                    [area_ranges for _ in range(num_imgs)],
+                    [use_legacy_coordinate for _ in range(num_imgs)], *args))
+        else:
+            tpfp = tpfp_fn(cls_dets[0], cls_gts[0], cls_gts_ignore[0], iou_thr,
+                           area_ranges, use_legacy_coordinate)
+            tpfp = [
+                tpfp,
+            ]
+
```
Comment (jbwang1997) [inline mmdet/core/evaluation/mean_ap.py:635]: Can merge all num_imgs > 1 branches together as below.

```
if num_imgs > 1:
    # compute tp and fp for each image with multiple processes
    assert nproc > 0, 'nproc must be at least one.'
    pool = Pool(min(nproc, num_imgs))
    tpfp = pool.starmap(
        tpfp_fn,
        zip(cls_dets, cls_gts, cls_gts_ignore,
             [iou_thr for _ in range(num_imgs)],
             [area_ranges for _ in range(num_imgs)],
             [use_legacy_coordinate for _ in range(num_imgs)], *args))
    pool.close()
else:
    tpfp = tpfp_fn(cls_dets[0], cls_gts[0], cls_gts_ignore[0], iou_thr,
                   area_ranges, use_legacy_coordinate)
    tpfp = [
        tpfp,
    ]
```

File: tools/analysis_tools/analyze_results.py [hunk 2]
```diff
@@ -39,11 +43,24 @@ def bbox_map_eval(det_result, annotation):
...
+        })
+        processes.append(p)
+
+    workers.close()
+    workers.join()
+
+    mean_aps = []
+    for p in processes:
+        mean_aps.append(p.get()[0])
+
```
Comment (jbwang1997) [inline tools/analysis_tools/analyze_results.py:63]: Can use pool.starmap instead.

Comment (chhluo) [reply tools/analysis_tools/analyze_results.py:63]: Passing parameters is cumbersome for pool.starmap here.
The result of 'apply_async' is out of order but more faster than `apply`.

**Review by jbwang1997 — APPROVED** [review]
LGTM

File: mmdet/core/evaluation/mean_ap.py (hunk 1 above)
Comment (chhluo) [reply mmdet/core/evaluation/mean_ap.py]: The code style looks good, but creating process pool in the loop is of low efficiency and unnecessary. I test this.

image

So, I think the original way is better.

**Comment by ZwwWayne** [comment]
Should try to add more unit tests to make sure that the AP calculation is not changed after modification.

**Pull request merged**